### PR TITLE
[HOLD] SHARE subject serialization [OSF-7788]

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
OSF-side of https://github.com/CenterForOpenScience/SHARE/pull/653

## Changes
* Update `Subject` serialization process
* Include information about deleted subjects when a `preprint.subjects` M2M relation changes

## Side effects
None expected. 

## Note
Tagged `[HOLD]` because this shouldn't go in until SHARE's ready for it.

## Ticket
Extension of [OSF-7788](https://openscience.atlassian.net/browse/OSF-7788)

### CODE REVIEW

File: osf/models/preprint_service.py [hunk 1]
```diff
@@ -108,10 +108,11 @@ def get_subjects(self):
                 ret.append(subj_hierarchy)
         return ret
 
-    def set_subjects(self, preprint_subjects, auth, save=False):
+    def set_subjects(self, preprint_subjects, auth):
```
Comment (mfraezz) [inline osf/models/preprint_service.py]: `save` kwarg removed because this modifies a M2M relation -- the through table is updated whether `.save` is called or not, and `.save` must be called here to trigger `throughsubject` deletions on SHARE.

File: tests/test_preprints.py [hunk 2]
```diff
@@ -336,8 +336,16 @@ def test_format_preprint(self):
...
-        assert sorted(subject['name'] for subject in subjects) == [s.bepress_text for h in self.preprint.subject_hierarchy for s in h]
+        s_ids = [s['@id'] for s in subjects]
+        ts_ids = [ts['subject']['@id'] for ts in through_subjects]
+        cs_ids = list(set(s['central_synonym']['@id'] for s in subjects))
+        for ts in ts_ids:
+            assert ts in s_ids
+            assert ts not in cs_ids  # Only aliased subjects are connected to self.preprint
+        for s in subjects:
+            subject = Subject.objects.get(text=s['name'])
+            assert s['uri'].endswith('v2/taxonomies/{}/'.format(subject._id))  # This cannot change
```
Comment (mfraezz) [inline tests/test_preprints.py]: Hard-coded here, rather than using `.absolute_api_v2_url`, to assert that it doesn't change. SHARE uses it for deduplication.

File: website/util/share.py [hunk 3]
```diff
@@ -67,3 +67,18 @@ def format_contributor(preprint, user, bibliographic, index):
...
+    if subject.id in context:
+        return context[subject.id]
+    context[subject.id] = GraphNode(
+        'subject',
+        name=subject.text,
+        is_deleted=False,
+        uri=subject.absolute_api_v2_url,
+    )
+    context[subject.id].attrs['parent'] = format_subject(subject.parent, context)
+    context[subject.id].attrs['central_synonym'] = format_subject(subject.central_alias, context)
```
Comment (chrisseto) [inline website/util/share.py]: As discussed over flowdock, `central_synonym` should be `None` for BePress values

File: website/util/share.py [hunk 4]
```diff
@@ -67,3 +67,18 @@ def format_contributor(preprint, user, bibliographic, index):
         creative_work=preprint,
         cited_as=user.fullname,
     )
+
+def format_subject(subject, context={}):
```
Comment (sloria) [inline website/util/share.py]: Drive-by comment: Don't use dicts (or any other mutable values) as default arguments.

File: website/preprints/tasks.py [hunk 5]
```diff
@@ -8,15 +8,15 @@
...
+from website.util.share import GraphNode, format_contributor, format_subject
 
 from website.identifiers.utils import request_identifiers_from_ezid, get_ezid_client, build_ezid_metadata, parse_identifiers
 
 logger = logging.getLogger(__name__)
 
 
 @celery_app.task(ignore_results=True)
-def on_preprint_updated(preprint_id, update_share=True):
+def on_preprint_updated(preprint_id, update_share=True, old_subjects=[]):
```
Comment (sloria) [inline website/preprints/tasks.py]: Same here: avoid mutable default arguments.

http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments
